lib/DxilValidation/DxilValidation.cpp

```diff
@@ -970,6 +970,175 @@ static void ValidateImmOperandForMathDxilOp(CallInst *CI, DXIL::OpCode opcode,
   }
 }
 
+static bool CheckFromRegisterInterpretations(uint32_t Ri) {
+  std::set<DXIL::ComponentType> ValidSet = {
+      DXIL::ComponentType::I16,         DXIL::ComponentType::U16,
+      DXIL::ComponentType::I32,         DXIL::ComponentType::U32,
+      DXIL::ComponentType::F16,         DXIL::ComponentType::F32,
+      DXIL::ComponentType::PackedS8x32, DXIL::ComponentType::PackedU8x32,
+      DXIL::ComponentType::U8,          DXIL::ComponentType::I8,
+      DXIL::ComponentType::F8_E4M3,     DXIL::ComponentType::F8_E5M2};
+
+  if (ValidSet.find(static_cast<DXIL::ComponentType>(Ri)) != ValidSet.end()) {
+    return true;
+  }
+  return false;
+}
+
+static bool CheckInMemoryInterpretations(uint32_t Mi) {
+  std::set<DXIL::ComponentType> ValidSet = {
+      DXIL::ComponentType::I16,     DXIL::ComponentType::U16,
+      DXIL::ComponentType::I32,     DXIL::ComponentType::U32,
+      DXIL::ComponentType::F16,     DXIL::ComponentType::F32,
+      DXIL::ComponentType::U8,      DXIL::ComponentType::I8,
+      DXIL::ComponentType::F8_E4M3, DXIL::ComponentType::F8_E5M2};
+
+  if (ValidSet.find(static_cast<DXIL::ComponentType>(Mi)) != ValidSet.end()) {
+    return true;
+  }
+  return false;
+}
+
+static bool CheckMatrixLayout(uint32_t Ml) {
+  std::set<DXIL::DXILMatrixLayout> ValidSet = {
+      DXIL::DXILMatrixLayout::RowMajor, DXIL::DXILMatrixLayout::ColumnMajor,
+      DXIL::DXILMatrixLayout::MulOptimal,
+      DXIL::DXILMatrixLayout::OuterProductOptimal};
+
+  if (ValidSet.find(static_cast<DXIL::DXILMatrixLayout>(Ml)) !=
+      ValidSet.end()) {
+    return true;
+  }
+  return false;
+}
+
+static void ValidateImmOperandsForMatVecOps(CallInst *CI, DXIL::OpCode opcode,
```

Comment:
Note - we should also do these sorts of checks in the builtins, not just in the validator.

Reply:
Yeah, can you open an issue for it, I'll handle it in a followup.

Reply:
#7329